src/neuralnet/cpid.cpp

```diff
@@ -0,0 +1,233 @@
+#include "neuralnet/cpid.h"
+#include "util.h"
+
+#include <boost/variant/apply_visitor.hpp>
+#include <openssl/md5.h>
+
+using namespace NN;
+
+namespace {
+//!
+//! \brief Gets the string representation of a mining ID object.
+//!
+struct MiningIdToStringVisitor : boost::static_visitor<std::string>
+{
+    //!
+    //! \brief Call the mining ID variant type's \c ToString() method directly.
+    //!
+    //! \param variant The object to create a string for.
+    //!
+    //! \return The string representation of the mining ID.
+    //!
+    template<typename T>
+    std::string operator()(const T& variant) const
+    {
+        return variant.ToString();
+    }
+};
+} // anonymous namespace
+
+// -----------------------------------------------------------------------------
+// Class: Cpid
+// -----------------------------------------------------------------------------
+
+Cpid::Cpid() : m_bytes({ })
+{
+}
+
+Cpid::Cpid(const std::vector<unsigned char>& bytes)
+{
+    if (bytes.size() == 16) {
+        std::copy_n(bytes.begin(), 16, m_bytes.begin());
+    } else {
+        m_bytes.fill(0x00);
+    }
+}
+
+Cpid Cpid::Parse(const std::string& hex)
+{
+    if (hex.size() != 32) {
+        return Cpid();
+    }
+
+    return Cpid(ParseHex(hex));
+}
+
+Cpid Cpid::Hash(const std::string& internal, const std::string& email)
+{
+    if (internal.empty() || email.empty()) {
+        return Cpid();
+    }
+
+    Cpid cpid;
+
+    // Even though the internal CPID is hex-encoded, BOINC creates the external
+    // CPID hash from the hex string, not from the encoded bytes:
+    //
+    std::vector<unsigned char> input(internal.begin(), internal.end());
+    input.insert(input.end(), email.begin(), email.end());
+
+    MD5(input.data(), input.size(), cpid.m_bytes.data());
+
+    return cpid;
+}
+
+bool Cpid::operator==(const Cpid& other) const
+{
+    return m_bytes == other.m_bytes;
+}
+
+bool Cpid::operator!=(const Cpid& other) const
+{
+    return m_bytes != other.m_bytes;
+}
+
+bool Cpid::IsZero() const
+{
+    const auto zero = [](const unsigned char& i) { return i == 0; };
+
+    return std::all_of(m_bytes.begin(), m_bytes.end(), zero);
+}
+
+bool Cpid::Matches(const std::string& internal, const std::string& email) const
+{
+    return m_bytes == Cpid::Hash(internal, email).m_bytes;
+}
+
+const std::array<unsigned char, 16>& Cpid::Raw() const
```

Comment:
Consider returning a copy here unless we want to use a reference for performance reason. It looks like the class is immutable so this should be safe, but the reference will be dangling if this object is used in a container which is then reorganized.

Reply:
Should we also change from an MD5 to sha256 hash for the CPID::Hash? Is it necessary for security? How much would it break things? Also, would have to be coded as a mandatory I think.

Reply:
@denravonska Ah, good point--let's use a copy. Only the tests use this method right now, so the overhead should be fine. Can I change that in the part 2 PR?

@jamescowens The `Hash()` method is used to generate an external CPID from an internal CPID and email address so that the wallet can validate the CPIDs loaded from client_state.xml locally to avoid running with a corrupt CPID. It replaces the obfuscated MD5 algorithm from the cpid.cpp class that I removed. BOINC's external CPIDs are MD5 hashes, so we can't change the format, but we won't use it in the protocol.

Reply:
@cyrossignol Sure :)

Reply:
@denravonska I remember why this returns a reference now--in the upcoming PR, I specialized `std::hash<T>` so that we can use `Cpid` objects as keys in `unordered_map`s (for lookup by CPID--beacons for example). A copy would happen there every time we access the maps by key if we go that route because it's calling this method to get the data to hash.

I'll do some reading about dangling references. In this case, `Raw()` is provided for low-level access which should happen only sparingly outside of the interface provided by the class.